<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Manual of GMIEC</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GMIEC-website</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="manual.html">Manual</a>
</li>
<li>
  <a href="faq.html">F.A.Q</a>
</li>
<li>
  <a href="contact.html">Contact</a>
</li>
<li>
  <a href="terms_of_use.html">Terms of use</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Manual of GMIEC</h1>

</div>


<div id="dependencies" class="section level4">
<h4><strong>Dependencies</strong></h4>
<p>GMIEC depends on several packages:</p>
<ul>
<li>bcp</li>
<li>ChIPpeakAnno</li>
<li>plyr</li>
<li>stats</li>
<li>shiny</li>
<li>shinydashboard</li>
<li>heatmaply</li>
<li>klaR</li>
<li>formattable</li>
<li>kableExtra</li>
<li>randomForest</li>
</ul>
</div>
<div id="how-to-install-gmiec" class="section level3">
<h3><strong>How to install GMIEC</strong></h3>
<p>Copy and paste the code inside the INSTALL.R on a R shell.</p>
<div id="how-to-run-gmiec-in-local" class="section level4">
<h4><strong>How to run GMIEC in local:</strong></h4>
<ol style="list-style-type: decimal">
<li>Go to <a href="https://github.com/guidmt/gmiec-shiny" class="uri">https://github.com/guidmt/gmiec-shiny</a></li>
<li>Download the repository or use git-clone</li>
<li>Go in the folder with the application and:
<ul>
<li>From console <strong>shinyAppDir(appDir=“/mydir/GMIEC_shiny”)</strong></li>
</ul></li>
</ol>
</div>
</div>
<div id="a-brief-summary" class="section level3">
<h3><strong>A brief summary</strong></h3>
<p>GMIEC consists of three application: - GMIEC-MD for the integration of multi-omics data - GMIEC-FD for the integration of only two data set - GMIEC-VIS to visualize the results of GMIEC-MD and GMIEC-FD</p>
<div id="step1-methods-of-analysis" class="section level4">
<h4><strong>Step1: Methods of analysis</strong></h4>
<p>GMIEC-MD implements two methods for the integration of the multi-omics data and the identification of groups of genes at the level of single patient. <strong>Method 1 (M1)</strong>, The first method (M1) it is based on an unsupervised machine learning approach, here we implemented a random forest algorithm from random-Forest R-package. For each patient is generate a matrix used as input for random forest that integrate the data and compute the similarities measures between the points (proximity values).The vector with the values of proximity is used as input for k-means to identify cluster of genes. Each cluster will contain genes that shared common genomic features. <strong>Method 2 (M2)</strong>,GMIEC-MD and GMIEC-FD implements also a logic approach to iden-tify (method 2, M2), in each patient, the genes that has specifics genomic features. For each patient (i) a matrix Mi = (g x r), in which n are genes and r rules is generated (see supplementary data). If a gene respect a giv-en rule is marked as [1], otherwise [0]. The clustering of the matrix M is performed with k-mode using a number cluster defined by the user.</p>
</div>
<div id="step2-compute-the-scores" class="section level4">
<h4><strong>Step2: Compute the scores</strong></h4>
<p>In the case that GMIEC-MD was run to use M1, a S-score like is computed for each module to measure if a group of genes are oncogenic or suppressed. In parallel to this, we identified also if a module is druggable or not. A module highly druggable is a module in which many genes are targeted by drugs. Finally we created a unique score to indicate the levels of alterations and druggability of a module. In the case that GMIEC is run with M2 or GMIEC-FD a level alteration module is compute (LAM score) to measure the level of genomic alteration. Similarly for M1, for each module an estimation between the number of genes links with drugs (gtg) and the total number of genes in module (tgm) is computed. Also in this case a combine scores is computed.</p>
</div>
<div id="annotation" class="section level4">
<h4><strong>Annotation</strong></h4>
<p>GMIEC-MD and GMIEC-FD can used as input genomic data such as ChIP-seq data (chromosome, start, end). In this case, the first step is the annotation of ChIP-seq data of a transcriptional factors are annotated using ChIPPeakAnno, considering a distance <strong>(d)</strong> of <strong>(n)</strong> bp. After this step the mean absolute deviation (MAD) of the annotated genes is estimated. Then, to filter the genes a bayesian analysis of change points of MAD is perfomed. This allowed to selected only the genes with high MAD values and without use a pre-determinate user threshold. In GMIEC this step is conducted by bcp function implemented in bcp R-package. This function implements the Bayesian change point methods develop by Wang and Emerson (2015). Then, after that the selection of genes is perfomed, for each patient start the core of the GMIEC analysis. The main aim of the workflow is the identification of specifics genes that shares the same genomic alterations (e.g. copy number variation, methylation, gene-expression changes, mutation etc.). In our context we will use the term “module” to define groups of genes that share the same genomics alterations and that are regulated by the same transcriptional factors. The identification of modules involves the integration of all omics data available.</p>
</div>
<div id="input-of-gmiec-md-and-gmiec-fd" class="section level4">
<h4><strong>Input of GMIEC-MD and GMIEC-FD:</strong></h4>
<ul>
<li><strong>omics data</strong>: gene-expression, copy-number epigenetics and mutations data in the form of tab-delimited files. In the case the genetic variants data, the user must create a file with only the mutations of interest (e.g. with deleterious variants).</li>
<li><strong>drugs-genes table</strong>: a tab delimited file with the association between genes and drugs. The user can create a custom file selecting only the drugs and genes that are important for the disease of interest.</li>
<li><strong>clinical data</strong>: a tab delimited file with clinical variables to include in the analysis.</li>
<li><strong>epigenetic data</strong> (not mandatory): a bed file with chromosome, start, end.</li>
<li><strong>list of genes</strong> (not mandatory): a list of genes of interest to consider for the analysis (e.g. genes of a pathway). The list of genes can be directly uploaded by the user. Otherwise if the user provides a file with genomic coordinates, the list of genes is obtained through a step of annotation.</li>
</ul>
</div>
<div id="output-of-gmiec-md-and-gmiec-fd" class="section level4">
<h4><strong>Output of GMIEC-MD and GMIEC-FD:</strong></h4>
<p>The output of GMIEC is a tab delimited file. This output contains the gene modules (GMs) and drug modules (DMs) identified during the analysis. The output reports also columns containing scores, that are a quantification of the levels of genomic alterations of the genes in GM, and the fraction of drugs links with a GMs. GMIEC allows also to gen-erate an html report (see figure S2) that can be interrogated directly click-ing on specific links. In particular, for each patient the groups of genes identified with common molecular features, the associated drugs and the clinical data provided by the user are reported.</p>
</div>
<div id="how-to-run-gmiec" class="section level4">
<h4><strong>How to run GMIEC:</strong></h4>
<p>If you use GMIEC-MD and GMIEC-FD two are the initial steps:</p>
<ol style="list-style-type: decimal">
<li>Upload all omics dataset</li>
<li><p>Upload the data with the association drug-gene</p></li>
<li><strong>Analysis considering only the annotated genes</strong>
<ul>
<li>Upload the bed file</li>
<li>Click on “Use only the annotated genes”</li>
<li>Select the number of clusters</li>
<li>Select if you want to run the analysis with “random + k-means (M1)”</li>
<li>Type the name for the output</li>
<li>Click on “Run Analysis”</li>
<li>When the analysis is finished click on “Download”</li>
</ul></li>
<li><strong>Analysis considering only the annotated genes and TF</strong>
<ul>
<li>Upload the bed file</li>
<li>Type the name of TF</li>
<li>Click on “Use only the annotated genes and explore the TF”</li>
<li>Select the number of clusters</li>
<li>The options “random + k-means (M1)” is not available in this case</li>
<li>Type the name for the output</li>
<li>Click on “Run Analysis”</li>
<li>When the analysis is finished click on “Download”</li>
</ul></li>
<li><strong>Analysis considering using a gene-list</strong>
<ul>
<li>Click on “Use a gene-list” (provide a tab-delimited file with one column reporting the genes)</li>
<li>Upload the gene-list
<ul>
<li>Click on “Use only the annotated genes and explore the TF”</li>
</ul></li>
<li>Select the number of clusters</li>
<li>Select if you want to run the analysis with “random + k-means (M1)”</li>
<li>Click on “Run Analysis”</li>
<li>When the analysis is finished click on “Download”</li>
</ul></li>
<li><strong>Analysis considering using all genes</strong>
<ul>
<li>Click on “Use all genes”</li>
<li>The options “random + k-means (M1)” is not available in this case</li>
<li>Click on “Run Analysis”</li>
<li>When the analysis is finished click on “Download”</li>
</ul></li>
</ol>
<ul>
<li><strong>Parameters for GMIEC</strong>
<ul>
<li>Click on “Use all genes”</li>
<li>Click on “Run Analysis”</li>
<li>When the analysis is finished click on “Download”</li>
</ul></li>
</ul>
</div>
<div id="how-to-run-gmiec-fd" class="section level4">
<h4><strong>How to run GMIEC-FD</strong></h4>
<ul>
<li>On the menu of GMIEC-Shiny, click on GMIC-FD and use the same procedures described for GMIEC-MD to analyze the data.</li>
</ul>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
