### GMIEC ###

GMIEC is a shiny application to integrate multi-omics data. GMIEC is a new user-friendly and flexible application that allows to integrate gene-expression, copy-number variation, methylation, Chromatin Immuno-Precipitation followed by sequencing (ChIP-seq) data, and mutations at the level of single patient. More interestingly, GMIEC facilitates the identification of groups of genes, that presents genomic alterations (e.g. copy number alterations).This can be a support help the researchers, to identify the de-regulated genes in a given disease, but also to identify discover novel pharmacological target molecules.

### Prerequisite ###

1. Install R >= 3.5.0, available on https://cran.r-project.org/ 
2. for windows user install RStudio (Optional), available on https://www.rstudio.com/products/rstudio/download/ 
dependencies R packages: shiny, stats,shinydashboard, rmarkdown, klaR, plyr, ChIPpeakAnno, stats, bcp, heatmaply, formattable, kableExtra, ggplot2, pysch, factoextra, randomForest.
To install all these packages, use the installation file provided on https://github.com/guidmt/GMIEC-shiny 

### See the wikipage with the manual with the details to run GMIEC ###

### Usage ###
The features of GMIEC make it a tool that could be useful in the context of the data integration, data mining and data-exploration at the level of single patient. GMIEC is a tool that is provided only to perform analysis, exploratory analysis in basic and bioinformatic research with a scientific usage. GMIEC was not develop as tool to use in the clinical practice and authors decline each responsibility on its usage.
